snowfall/training/asg_graph.py

```diff
@@ -0,0 +1,164 @@
+    '''
+    assert 0 not in phones
+    final_state = len(phones) + 1
+    rules = ''
+    for i in range(1, final_state):
+        rules += f'0 {i} {phones[i-1]} 0.0\n'
+
+    for i in range(1, final_state):
+        for j in range(1, final_state):
+            rules += f'{i} {j} {phones[j-1]} 0.0\n'
+        rules += f'{i} {final_state} -1 0.0\n'
+    rules += f'{final_state}'
+    return k2.Fsa.from_str(rules)
+
+
+class AsgTrainingGraphCompiler(object):
+
+    def __init__(self,
+                 L_inv: k2.Fsa,
+                 phones: k2.SymbolTable,
+                 words: k2.SymbolTable,
+                 oov: str = '<UNK>'):
+        '''
+        Args:
+          L_inv:
+            Its labels are words, while its aux_labels are phones.
+        phones:
+          The phone symbol table.
+        words:
+          The word symbol table.
+        oov:
+          Out of vocabulary word.
+        '''
+
+        if L_inv.properties & k2.fsa_properties.ARC_SORTED != 0:
+            L_inv = k2.arc_sort(L_inv)
+
+        assert L_inv.requires_grad is False
+
+        assert oov in words
+
+        self.L_inv = L_inv
+        self.phones = phones
+        self.words = words
+        self.oov = oov
+
+        phone_symbols = get_phone_symbols(phones)
+        phone_symbols_with_blank = [0] + phone_symbols
+
+        ctc_topo = build_ctc_topo(phone_symbols_with_blank)
+        assert ctc_topo.requires_grad is False
+
+        self.ctc_topo = ctc_topo
+
+    def compile(self, texts: Iterable[str],
+                P: k2.Fsa) -> Tuple[k2.Fsa, k2.Fsa]:
+        '''Create numerator and denominator graphs from transcripts
+        and the bigram phone LM.
+
+        Args:
+          texts:
+            A list of transcripts. Within a transcript, words are
+            separated by spaces.
+          P:
+            The bigram phone LM created by :func:`create_bigram_phone_lm`.
+        Returns:
+          A tuple (num_graph, den_graph), where
+
+            - `num_graph` is the numerator graph. It is an FsaVec with
+              shape `(len(texts), None, None)`.
+
+            - `den_graph` is the denominator grap. It is an FsaVec with the same
+              shape of the `num_graph`.
+        '''
+        assert P.is_cpu()
+
+        den = k2.intersect(self.ctc_topo, P).invert_()
+        den = k2.connect(den)
+
+        decoding_graphs = k2.create_fsa_vec(
+            [self.compile_one_and_cache(text) for text in texts])
+
+        num = k2.compose(den, decoding_graphs)
+        num = k2.connect(num)
+        num = k2.arc_sort(num)
+
+        den = k2.create_fsa_vec([den.detach()] * len(texts))
+
+        return num, den
+
+    @lru_cache(maxsize=100000)
+    def compile_one_and_cache(self, text: str) -> k2.Fsa:
+        '''Convert transcript to an Fsa.
+
+        Args:
+          text:
+            The transcript containing words separated by spaces.
+
+        Returns:
+          Return an FSA corresponding to the transcript.
```

Comment:
.. this documentation is a little vague (and I think you could have avoided a bug if you had tried to make it more precise).  You don't specify what the labels and aux_labels are, and whether it includes contributions from ctc_topo and P.

Reply:
Will make it clearer.
The returned FSA has `phones` as `labels`, and `epsilons` or `words` as `aux_labels`.

> I think you could have avoided a bug

Could you give a hint where the bug is?